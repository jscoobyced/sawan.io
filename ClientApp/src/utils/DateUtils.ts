export class DateUtil {
    public static toHumanHours(date: Date): string {
        return `${this.toDigits(date.getHours(), 2)}:${this.toDigits(date.getMinutes(), 2)}`;
    }

    public static dateToString(date: Date): string {
        const lDate = new Date(date);
        const year = lDate.getFullYear();
        const month = DateUtil.toDigits(lDate.getMonth() + 1, 2);
        const day = DateUtil.toDigits(lDate.getDate(), 2);
        return `${year}-${month}-${day}`;
    }

    public static dateToYyyyMmString(date: Date): string {
        const lDate = new Date(date);
        const year = lDate.getFullYear();
        const month = DateUtil.toDigits(lDate.getMonth() + 1, 2);
        return `${year}${month}`;
    }

    public static defaultDate(now = false): Date {
        if (now) {
            const nowDate = new Date();
            const month = DateUtil.toDigits(nowDate.getMonth() + 1, 2);
            const day = DateUtil.toDigits(nowDate.getDate(), 2);
            return new Date(`${nowDate.getFullYear()}-${month}-${day}T00:00:00Z`);
        }

        return new Date('2019-01-01T00:00:00Z');
    }

    private static toDigits(value: number, length: number): string {
        let result: string = value.toString();
        while (result.length < length) {
            result = "0" + result;
        }

        return result;
    }
}
