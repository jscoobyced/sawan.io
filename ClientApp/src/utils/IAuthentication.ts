export interface IAuthentication {
    getTokenId: () => string;
    init: (callBack: () => void) => void;
    isAuthSupported: () => boolean;
    signOut: () => Promise<any>;
    renderButton: (onSignIn: () => void) => void;
}
