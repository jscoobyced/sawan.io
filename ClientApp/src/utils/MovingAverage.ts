export class MovingAverage {
  public ma(data: number[], size?: number): number[] {
    if (data === undefined || data.length === 0) {
      return [];
    }

    const isNumber = (subject: any) => typeof subject === 'number';
    const { length } = data;
    const ret: number[] = [];

    if (!size) {
      ret[0] = data.reduce((a, b) => a + b) / length;
      return ret;
    }

    if (size <= 1) {
      return data.slice();
    }

    if (size > length) {
      return Array(length);
    }

    const prepare = size - 1;
    let sum = 0;
    let i = 0;
    let counter = 0;
    let datum;

    for (; i < length && counter < prepare; i += 1) {
      datum = data[i];

      if (isNumber(datum)) {
        sum += datum;
        counter += 1;
      }
    }

    for (; i < length; i += 1) {
      datum = data[i];

      if (isNumber(datum)) {
        sum += datum;
      }

      if (isNumber(data[i - size])) {
        sum -= data[i - size];
      }

      ret[i] = sum / size;
    }

    return ret;
  }
}
