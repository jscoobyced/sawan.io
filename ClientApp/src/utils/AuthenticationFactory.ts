import { IAuthentication } from "./IAuthentication";

export class AuthenticationFactory {
    public static getAuthentication(): IAuthentication {
        return AuthenticationFactory.authentication;
    }

    public static registerAuthentication(authentication: IAuthentication, force = false) {
        if (!AuthenticationFactory.authentication || force) {
            AuthenticationFactory.authentication = authentication;
        }
    }

    private static authentication: IAuthentication;
}
