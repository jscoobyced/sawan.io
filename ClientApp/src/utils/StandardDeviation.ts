import { MovingAverage } from './MovingAverage';

export class StandardDeviation {
  public sd(data: number[], size?: number): number[] {
    const safeSize = size === undefined ? 1 : size;
    const { length } = data;
    const avg = new MovingAverage().ma(data, safeSize);
    const ret: number[] = [];

    let i = safeSize - 1;
    let j;
    let sum;

    for (; i < length; i += 1) {
      sum = 0;
      j = i - safeSize + 1;

      for (; j <= i; j += 1) {
        sum += (data[j] - avg[i]) ** 2;
      }

      ret[i] = Math.sqrt(sum / safeSize);
    }

    return ret;
  }
}
