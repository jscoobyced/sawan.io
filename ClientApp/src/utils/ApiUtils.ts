import { AuthenticationFactory } from "./AuthenticationFactory";

export class ApiUtils {
    public static async fetchData<T>(url: string): Promise<T> {
        const response = await fetch(url);
        if (!response.ok) {
            throw new Error(response.statusText);
        }
        return response.json() as Promise<T>;
    }

    public static async postData<T, U>(url: string, data: T): Promise<U> {
        const user = AuthenticationFactory.getAuthentication().getAuthenticatedUser();
        let token = "";
        if (user && user.token) {
            token = user.token;
        }
        return fetch(url, {
            method: 'POST',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json',
                'Authorization': 'Bearer ' + token
            },
            body: JSON.stringify({
                data,
                accessToken: AuthenticationFactory.getAuthentication().getTokenId()
            })
        })
            .then(response => {
                return response.json();
            }).then(response => {
                return response;
            });
    }
}
