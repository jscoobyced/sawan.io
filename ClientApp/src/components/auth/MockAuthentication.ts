import { IAuthentication } from "../../utils/IAuthentication";

export class MockAuthentication implements IAuthentication {
    public getAuthenticatedUser = () => {
        return {
            email: "user@example.org",
            family_name: "Smith",
            given_name: "Kevin",
            group: "Admin",
            id: "123456789",
            locale: "en_US",
            name: "Kevin Smith",
            picture: "",
            token: "",
            verified_email: true
        };
    }

    public getTokenId = (): string => {
        return "123456789";
    }
    public init = () => {
        return;
    }

    public isReady = (): boolean => {
        return true;
    }

    public isAuthSupported = (): boolean => {
        return true;
    }

    public signOut = (): Promise<any> => {
        return Promise.resolve();
    }

    public renderButton = (onSignIn: () => void) => {
        gapi.signin2.render('g-signin2', {
            scope: 'profile email',
            longtitle: false,
            theme: 'dark',
            onsuccess: onSignIn,
        });
    }
    public isAdmin() {
        return true;
    }
}
