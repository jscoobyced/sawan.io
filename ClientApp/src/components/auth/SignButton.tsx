import React from 'react';
import { AuthenticationFactory } from '../../utils/AuthenticationFactory';

export interface SignButtonState {
    isSignedIn: boolean;
    numberOfAttemtps: number;
}

export class SignButton extends React.Component<{}, SignButtonState> {

    public constructor(props: {}, state: SignButtonState) {
        super(props, state);
        this.state = {
            isSignedIn: false,
            numberOfAttemtps: 0
        };
        AuthenticationFactory.getAuthentication().init(this.onSignIn);
    }

    public render() {
        return this.state.isSignedIn
            ? <a href='#' onClick={this.signOut} >Sign-Out</a>
            : <div id='g-signin2'></div>;
    }

    private readonly onSignIn = () => {
        if (AuthenticationFactory.getAuthentication().isAuthSupported()) {
            this.setState({
                isSignedIn: true
            });
        }
    }

    private readonly signOut = () => {
        AuthenticationFactory.getAuthentication().signOut().then(() => {
            this.setState({
                isSignedIn: false
            });
            AuthenticationFactory.getAuthentication().renderButton(this.onSignIn);
        });
    }
}
