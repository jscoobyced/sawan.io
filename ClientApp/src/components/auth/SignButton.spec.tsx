import { shallow } from 'enzyme';
import React from 'react';
import { AuthenticationFactory } from '../../utils/AuthenticationFactory';
import { MockAuthentication } from './MockAuthentication';
import { SignButton } from './SignButton';

test('SignButton default component is unchanged.', () => {
    AuthenticationFactory.registerAuthentication(new MockAuthentication());
    const signButton = shallow(<SignButton />);
    expect(signButton).toMatchSnapshot();
});

test('SignButton with authentication ready.', () => {
    const mockAuthentication = new MockAuthentication();
    mockAuthentication.init = jest.fn().mockImplementation();
    AuthenticationFactory.registerAuthentication(mockAuthentication, true);
    const signButton = shallow(<SignButton />);
    expect(signButton).toMatchSnapshot();
    expect(mockAuthentication.init).toBeCalledTimes(1);
});
