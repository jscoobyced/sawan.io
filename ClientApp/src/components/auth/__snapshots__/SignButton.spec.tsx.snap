// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignButton default component is unchanged. 1`] = `
<div
  id="g-signin2"
/>
`;

exports[`SignButton with authentication ready. 1`] = `
<div
  id="g-signin2"
/>
`;
